beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,722 @@
+                validator_index,
+                slot: sync_signature.slot,
+            });
+        }
+
+        // The aggregate signature of the sync committee message is valid.
+        verify_sync_signature(chain, &sync_signature)?;
+
+        // Now that the sync committee message has been fully verified, store that we have received a valid
+        // sync committee message from this validator.
+        //
+        // It's important to double check that the sync committee message still hasn't been observed, since
+        // there can be a race-condition if we receive two sync committee messages at the same time and
+        // process them in different threads.
+        if chain
+            .observed_sync_contributors
+            .write()
+            .observe_validator(sync_signature.slot, validator_index as usize)
+            .map_err(BeaconChainError::from)?
+        {
+            return Err(Error::PriorSyncSignatureKnown {
+                validator_index,
+                slot: sync_signature.slot,
+            });
+        }
+
+        Ok(Self {
+            sync_signature,
+            subnet_positions,
+        })
+    }
+
+    /// A helper function to add this sync committee message to `beacon_chain.naive_sync_aggregation_pool`.
+    pub fn add_to_pool<T: BeaconChainTypes>(self, chain: &BeaconChain<T>) -> Result<Self, Error> {
+        chain.add_to_naive_sync_aggregation_pool(self)
+    }
+
+    /// Returns the subcommittee positions for the sync signature, keyed on the `SyncSubnetId` for
+    /// the subnets the signature should be sent on.
+    pub fn subnet_positions(&self) -> HashMap<SyncSubnetId, Vec<usize>> {
+        self.subnet_positions.clone()
+    }
+
+    /// Returns the wrapped `SyncCommitteeSignature`.
+    pub fn sync_signature(&self) -> &SyncCommitteeSignature {
+        &self.sync_signature
+    }
+}
+
+/// Returns `Ok(())` if the `beacon_block_root` is known to this chain.
+///
+/// The block root may not be known for two reasons:
+///
+/// 1. The block has never been verified by our application.
+/// 2. The block is prior to the latest finalized block.
+///
+/// Case (1) is the exact thing we're trying to detect. However case (2) is a little different, but
+/// it's still fine to reject here because there's no need for us to handle sync committee messages that are
+/// already finalized.
+fn verify_head_block_is_known<T: BeaconChainTypes>(
+    chain: &BeaconChain<T>,
+    beacon_block_root: Hash256,
+) -> Result<ProtoBlock, Error> {
+    if let Some(block) = chain.fork_choice.read().get_block(&beacon_block_root) {
+        Ok(block)
+    } else {
+        Err(Error::UnknownHeadBlock { beacon_block_root })
+    }
+}
+
+/// Verify that the `sync_contribution` is within the acceptable gossip propagation range, with reference
+/// to the current slot of the `chain`.
+///
+/// Accounts for `MAXIMUM_GOSSIP_CLOCK_DISPARITY`.
+pub fn verify_propagation_slot_range<T: BeaconChainTypes, E: SlotData>(
```

Comment:
Would prefer a name other than `E` for this type-param (because `E` is usually `EthSpec`).